enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,320 @@
+ICC is currently not capable of building a *mimimal ISO*, i.e. a CoreOS ISO
+without the root filesystems. While not a hard blocker of this work, the
+minimal ISO support should be implemented because
+- it improves compatibility with BMCs that do not accept large ISOs,
+- it reduces the resource footprint on the Ironic side.
+
+#### Notes
+
+- The new annotation with accept a URI instead of a full Ignition configuration
+  or a link to a secret because the configuration can be quite large when
+  many hosts are present.
+
+- ICC will replicate the graceful wait period logic from the assisted service:
+  the image will be built not earlier than 60 seconds from the last change
+  to the `PreprovisioningImage` object. This is done for two reasons: to give
+  the user enough time to apply all changes that may affect the build and to
+  reduce the probability of a race between BMAC and ICC.
+
+- Currently, BMO never deletes `PreprovisioningImage` resources. Since all
+  Ignition configurations are stored in memory, it may cause a significant
+  memory usage in the ZTP case. We will need to fix BMO to delete
+  `PreprovisioningImage` resources when the `BareMetalHost` becomes
+  `provisioned` or is deleted.
+
+- The same process will be applied to `initrd` images for (i)PXE booting.
+
+### Risks and Mitigations
+
+- More containers will be downloaded on the ramdisk start-up. We'll need to
+  account for more RAM and a longer start-up time.
+
+- Users will be responsible for correctly linking `BareMetalHost`s with
+  `InfraEnv` resources.
+
+## Design Details
+
+### Open Questions
+
+- The assisted mode of operation is to boot the assisted agent and wait for
+  instructions from the service (and indirectly - from the user). During this
+  time, the `BareMetalHost` state will be `provisioning`, and timeouts may
+  apply. We may need a way to tell Ironic to never time out this operation.
+
+### Test Plan
+
+- Building the new ISO will be covered by the existing ZTP e2e tests.
+- A new ZTP job will need to be created for network booting.
+
+### Graduation Criteria
+
+This work will be GA immediately since there is no phasing possible or planned.
+Only ZTP will be affected. The new behavior will be fully opt-in from the
+assisted service's (more specifically, BMAC) point of view.
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+This change does not introduce new resources, so upgrading or downgrading them
+is not a concern.
+
+For the new flow to apply, it must be explicitly requested through the
+`BareMetalHost` API. Otherwise, the old flow is used.
+
+There are no plans to remove the code that supports the *live ISO* flow from
+Ironic, Metal3 or OpenShift.
+
+### Version Skew Strategy
+
+The image-customization-controller will need to be updated before the new flow
+will be usable. Given that ZTP releases trail OpenShift releases (and are tied
```

Comment:
We'll have to allow ZTP and OCP to be upgraded independently and so ZTP will have to account for different versions of BMO with different expectations. How will it do that with no API version increment for BMH?

Reply:
I don't know if we've ever had to deal with this before, have we? ZTP needed some changes, but it just didn't work until the version of OCP that had those changes.

Reply:
> I don't know if we've ever had to deal with this before, have we? 

Well, at some point we introduced the live ISO support itself. I don't think BMAC actually checks the schema to validate the support is there.

Reply:
Did we just assume the new BMAC would use the new BMO?

If we're introducing a behavior without an API version update because that API is an annotation and the CRD is owned upstream, then I don't see a safer way to check for the behavior than looking at the version of something. The brute force way would be to check the ClusterVersion resource for the OCP version. Something more subtle might involve having one of the operators drop an annotation on a configuration resource when the behavior is available. Perhaps someone on the API team can suggest alternatives based on their experience.

Reply:
We could have the CBO put something in the status of the Provisioning CR.
Or even just checking the CBO version with CVO (rather than the cluster version) would be an improvement.